libclamav/udf.c

```diff
@@ -0,0 +1,838 @@
+    }
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, NUM_GENERIC_VOLUME_DESCRIPTORS * VOLUME_DESCRIPTOR_SIZE);
+    if (NULL == buffer) {
+        ret = CL_SUCCESS;
+        goto done;
+    }
+
+    for (i = 0; i < NUM_GENERIC_VOLUME_DESCRIPTORS; i++) {
+        gvsd       = (GenericVolumeStructureDescriptor *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
+        lastOffset = idx;
+
+        if (strncmp("BEA01", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("BOOT2", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("CD001", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("CDW02", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("NSR02", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("NSR03", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("TEA01", gvsd->standardIdentifier, 5)) {
+        } else {
+            cli_warnmsg("Unknown Standard Identifier '%s'\n", gvsd->standardIdentifier);
+            break;
+        }
+
+        idx += VOLUME_DESCRIPTOR_SIZE;
+    }
+
+    memset(&fileIdentifierList, 0, sizeof(PointerList));
+    memset(&fileEntryList, 0, sizeof(PointerList));
+
+    while (1) {
+
+        if (!isInitialized) {
+
+            if (CL_SUCCESS != (ret = initPointerList(&fileIdentifierList))) {
+                goto done;
+            }
+
+            if (CL_SUCCESS != (ret = initPointerList(&fileEntryList))) {
+                goto done;
+            }
+
+            if (NULL == (pvd = getPrimaryVolumeDescriptor(ctx, &idx, &lastOffset))) {
+                goto done;
+            }
+
+            if (NULL == (iuvd = getImplementationUseVolumeDescriptor(ctx, &idx, &lastOffset))) {
+                goto done;
+            }
+
+            if (NULL == (lvd = getLogicalVolumeDescriptor(ctx, &idx, &lastOffset))) {
+                goto done;
+            }
+
+            if (NULL == (pd = getPartitionDescriptor(ctx, &idx, &lastOffset))) {
+                goto done;
+            }
+
+            if (NULL == (usd = getUnallocatedSpaceDescriptor(ctx, &idx, &lastOffset))) {
+                goto done;
+            }
+
+            if (NULL == (td = getTerminatingDescriptor(ctx, &idx, &lastOffset))) {
+                goto done;
+            }
+
+            /*May not be every file, need to verify.*/
+            if (NULL == (lvid = getLogicalVolumeIntegrityDescriptor(ctx, &idx, &lastOffset))) {
+                goto done;
+            }
+
+            if (NULL == (td = getTerminatingDescriptor(ctx, &idx, &lastOffset))) {
+                goto done;
+            }
+
+            if (NULL == (avdp = getAnchorVolumeDescriptorPointer(ctx, &idx, &lastOffset))) {
+                goto done;
+            }
+
+            if (NULL == getFileSetDescriptor(ctx, &idx, &lastOffset)) {
+                goto done;
+            }
+
+            isInitialized = true;
+        }
+
+        buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
```

Comment:
I see you have a loop at the bottom to unneed all the offsets that you (probably) needed here.  Instead you could use `fmap_need_off_once()` so you don't have to unneed them, or else unneed before each `goto done;`  and at the end of the loop.  Not ... great, but maybe better than accidentally unneeding something that was never needed ( not actually sure if there is a consequence to that ).

Reply:
pretty sure there is no consequence to that.  I am also hesitant to clean it up each time, I am returning pointers from specific functions.  If there is no consequence, I am inclined to unneed everything at the end to avoid missing one.  I am open to discussion on that, though.

Reply:
Realistically I don't think there are any consequences.  The unneed code winds up here:
```c
static void fmap_unneed_page(fmap_t *m, uint64_t page)
{
    uint64_t s = fmap_bitmap[page];

    if ((s & (FM_MASK_PAGED | FM_MASK_LOCKED)) == (FM_MASK_PAGED | FM_MASK_LOCKED)) {
        /* page is paged and locked: check lock count */
        s &= FM_MASK_COUNT;
        if (s > 1) /* locked more than once: dec lock count */
            fmap_bitmap[page]--;
        else if (s == 1) /* only one lock left: unlock and begin aging */
            fmap_bitmap[page] = FM_MASK_COUNT | FM_MASK_PAGED;
        else
            cli_errmsg("fmap_unneed: inconsistent map state\n");
        return;
    }
    cli_warnmsg("fmap_unneed: unneed on a unlocked page\n");
    return;
}
```
ref: https://github.com/micahsnyder/clamav-micah/blob/89cd0df3d7d4c2163d79056b5c0ec55c4992c8f4/libclamav/fmap.c#L678-L695

So what you're doing is probably fine.